### PR TITLE
Genscript new changes

### PR DESCRIPTION
**Pull request by nbhoski**
Changes include the new Genscript options 

For Freestyle Project 

* Use Parallel : bolean 
* Strict : Boolean
* Logging level : String <Selection Box>
         * Default
         * None
         * Terse
         * Concise
         * Detailed
         * Verbose 
* Output Detail : String <Selection box>
         * Default
         * None
         * Terse
         * Concise
         * Detailed
         * Verbose 

For DSL 

`useParallel` : boolean 
`strict` : boolean 
`loggingLevel` : String 
         * Default
         * None
         * Terse
         * Concise
         * Detailed
         * Verbose 
`outputDetail` : String 
         * Default
         * None
         * Terse
         * Concise
         * Detailed
         * Verbose

### CODE REVIEW

File: src/test/java/com/mathworks/ci/RunMatlabTestsBuilderTest.java [hunk 1]
```diff
@@ -524,4 +524,135 @@ public void verifySystemTempDirDeleted() throws Exception {
...
+     * Test To verify if Logging level is set correctly 
+     * 
+     */
+
+    @Test
+    public void verifyLoggingLevelSet() throws Exception {
+        this.buildWrapper.setMatlabBuildWrapperContent(new MatlabBuildWrapperContent(
+                Message.getValue("matlab.custom.location"), getMatlabroot("R2018b")));
+        project.getBuildWrappersList().add(this.buildWrapper);
+        Map<String, String> logginggLevel = new HashMap<String, String>();
```
Comment (davidbuzinski) [inline src/test/java/com/mathworks/ci/RunMatlabTestsBuilderTest.java:538]: Should `logginggLevel ` be `loggingLevel`? (Extra g)

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select to fail the test suite on warnnings</p>
```
Comment (davidbuzinski) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html:2]: Spelling: warnings

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select the logging level option from the list</p>
```
Comment (davidbuzinski) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html:2]: For some reason, I don't like the "option from the list" part here. It doesn't seem like the other drop downs have it. Houman would probably have a good suggestion.

**Review by davidbuzinski — COMMENTED** [review]
Nice! I like the changes. 

Side note: why is the CI build failing? It looks like the actual Maven build is successful.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly [hunk 4]
```diff
@@ -1,22 +1,45 @@
...
-   <f:optionalProperty field="selectByFolder" title="By folder name" />  
+	<f:entry field="useParallel" title="Use parallel " checked="${instance.useParallel}">
+	    <f:checkbox/>
+	</f:entry>
+
+	<f:entry field="strict" title="Strict " checked="${instance.strict}">
+	    <f:checkbox/> 
+	</f:entry> 
+
+	<!-- Verbocity fields -->
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly:19]:
```suggestion
	<!-- Verbosity fields --> 
```

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html (hunk 3 above)
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html:2]:
```suggestion
    <p>Maximum verbosity level for logged diagnostics included for the test run.</p>  
```

File: src/main/java/com/mathworks/ci/RunMatlabTestsStep.java [hunk 5]
```diff
@@ -155,9 +195,11 @@ private String getInputArgs() {
         inputArgs.add("'Test'");
 
         args.forEach((key, val) -> {
-            if(key.equals("SourceFolder") || key.equals("SelectByFolder") && val != null){
+            if (key.equals("SourceFolder") || key.equals("SelectByFolder")
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:198]: some genscript parameters are not passed with single quotes eg :genscript(sourceFolder, sourceFolder[], PDFTestArtifact, 'PDFTestArtifact')

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-outputDetail.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select verbosity level of output details for test results in the log</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-outputDetail.html:2]:
```suggestion
    <p>Display level for event details output by the test run.</p>  
```

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html (hunk 2 above)
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html:2]:
```suggestion
    <p>Whether to apply strict checks when running the tests. For example, the framework generates a qualification failure if a test issues a warning.</p>  
```

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-useParallel.html [hunk 7]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select to run the tests in parallel</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-useParallel.html:2]: Whether to run tests in parallel. This feature requires a Parallel Computing Toolbox license, and might not be compatible with other options. If incompatible, testing occurs in serial regardless of the value of this feature.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select the logging level option from the list</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html:2]: See suggestions above

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-outputDetail.html [hunk 9]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select verbosity level of output details for test results in the log</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-outputDetail.html:2]: See suggestions above

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select to fail the test suite on warnnings</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html:2]: See suggestions above

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html [hunk 11]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select to run the tests in parallel</p>
```
Comment (mcafaro) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html:2]: See suggestions above

File: src/main/java/com/mathworks/ci/RunMatlabTestsStep.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:198]: oops my Bad i got your point.

File: src/main/java/com/mathworks/ci/RunMatlabTestsStep.java [hunk 12]
```diff
@@ -174,6 +216,10 @@ private Map<String, String> getGenscriptArgs() {
         args.put("CoberturaCodeCoverage", getCodeCoverageCobertura());
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:216]: The constants what we have are for parameter Names and not for parameter values for eg 
genscript ('TAPTestResults', 'path/to/some/folder') what you are seeing in constants are the first part of the parameter which are enclosed in single quotes. Values may or may not be in quotes. In this case all 4 values are first part of parameter hence i added them in constants However i see your point here i can make it look consistent without using MATLAB constants.

File: src/test/java/com/mathworks/ci/RunMatlabTestsBuilderTest.java (hunk 1 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabTestsBuilderTest.java:538]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html (hunk 2 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html:2]: Done.

File: src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java [hunk 13]
```diff
@@ -347,6 +416,24 @@ private String getInputArguments() {
...
         }
+        
+        // Add Logging level
+        if (!getLoggingLevel().equalsIgnoreCase("default")) {
+            inputArgsList.add("'" + MatlabBuilderConstants.LOGGING_LEVEL + "'" + "," + "'"
+                    + getLoggingLevel() + "'");
+        }
+
+        // Add Logging level
+        if (!getLoggingLevel().equalsIgnoreCase("default")) {
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsBuilder.java:427]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly [hunk 14]
```diff
@@ -1,22 +1,45 @@
...
+	</f:entry> 
+
+	<!-- Verbocity fields --> 
+	<f:entry name="loggingLevel " title="Logging level" field="loggingLevel" default="Default">
+	    <f:select/>
+	</f:entry>
+	
+	<f:entry name="outputDetail" title="Output detail " field="outputDetail" default="Default">
+	    <f:select/>
+	</f:entry>
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly:26]: Yes i am open to do that

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-outputDetail.html (hunk 6 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-outputDetail.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-useParallel.html (hunk 7 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-useParallel.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html (hunk 8 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-outputDetail.html (hunk 9 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-outputDetail.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html (hunk 2 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html (hunk 11 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html (hunk 10 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html:2]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html (hunk 8 above)
Comment (mcafaro) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-loggingLevel.html:2]: Don't see change

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly (hunk 4 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/config.jelly]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html (hunk 3 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-loggingLevel.html]: Done.

File: src/main/java/com/mathworks/ci/RunMatlabTestsStep.java [hunk 15]
```diff
@@ -155,9 +195,11 @@ private String getInputArgs() {
         inputArgs.add("'Test'");
 
         args.forEach((key, val) -> {
-            if(key.equals("SourceFolder") || key.equals("SelectByFolder") && val != null){
+            if (key.equals("SourceFolder") || key.equals("SelectByFolder")
+                    || key.equals("UseParallel")
+                    || key.equals("Strict") && val != null) {
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:200]: Is this intended to work as written or as
```java
            if ((key.equals("SourceFolder") || key.equals("SelectByFolder")
                    || key.equals("UseParallel")
                    || key.equals("Strict")) && val != null) {
```
I think `&&` gets evaluated before `||` in java so having `val == nul` might only compare with the `key.equals("Strict")` part of this expression.

Comment (mcafaro) [reply src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:200]: If you see the same condition being evaluated for more than one branch of an if/else, it usually needs a refactor. In this case, you could pull out the `val != null` as a separate block.

```java
if (val == null) {
    continue;
}
if (key...) {
} else {
}
```
OR
```java
if (val != null) {
    if (key...) {
    } else {
    }
}
```

Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabTestsStep.java:200]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html [hunk 16]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Select to fail the test suite on warnings</p>
```
Comment (davidbuzinski) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html:2]: This looks like the only one without a period at the end.

**Review by davidbuzinski — APPROVED** [review]

File: src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html (hunk 16 above)
Comment (nbhoski) [reply src/main/resources/com/mathworks/ci/RunMatlabTestsBuilder/help-strict.html]: Done.

**Review by davidbuzinski — APPROVED** [review]

**Review by mcafaro — APPROVED** [review]

File: src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java [hunk 17]
```diff
@@ -198,4 +200,102 @@ public void verifyTestSelectByTag () throws Exception {
         j.assertLogContains("myTestTag", build);
         j.assertBuildStatusSuccess(build);
     }
+    
+    /*@Integ
+     * Verify outPutDetail set
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java:205]: Done.

File: src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java [hunk 18]
```diff
@@ -198,4 +200,102 @@ public void verifyTestSelectByTag () throws Exception {
...
+                e.printStackTrace();
+            }
+        });
+    }
+    
+    /*@Integ
+     * Verify when useParallel Set 
+     */
+    
+    public void verifyRunParallelSet () throws Exception {
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java:262]: Done.

File: src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java [hunk 19]
```diff
@@ -198,4 +200,102 @@ public void verifyTestSelectByTag () throws Exception {
...
+                "node {runMATLABTests(useParallel: true)}", true));
+        WorkflowRun build = project.scheduleBuild2(0).get();
+        j.assertLogContains("runInParallel", build);
+    }
+    
+    /*@Integ
+     * Verify when useParallel Not Set 
+     */
+    
+    public void verifyRunParallelNotSet () throws Exception {
```
[the remark this answers is not in the document]
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabTestsStepTest.java:273]: Done.

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html [hunk 20]
```diff
@@ -0,0 +1,4 @@
+<div>
+    <p>Whether to run tests in parallel. This feature requires a Parallel Computing Toolbox license, and might not be 
+      compatible with other options. If incompatible, testing occurs in serial regardless of the value of this feature.</p>
```
Comment (mw-hrastega) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-useParallel.html:3]: Consider replacing "regardless of the value of this feature." with "regardless of the specified value."

File: src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html [hunk 21]
```diff
@@ -0,0 +1,3 @@
+<div>
+    <p>Whether to apply strict checks when running the tests. For example, the framework generates a qualification failure if a test issues a warning.</p>
```
Comment (mw-hrastega) [inline src/main/resources/com/mathworks/ci/RunMatlabTestsStep/help-strict.html:2]: framework or plugin?

**Review by mw-hrastega — COMMENTED** [review]
Suggesting some minor edits.

**Pull request merged**